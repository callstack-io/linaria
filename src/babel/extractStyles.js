/* @flow */

import type { BabelTaggedTemplateExpression, State } from './types';

export default function extractStyles(
  taggedTemplateExpression: BabelTaggedTemplateExpression,
  state: State
) {}
