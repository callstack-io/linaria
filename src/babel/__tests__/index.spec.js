jest.mock('../extractStyles', () => ({ __esModule: true, default: jest.fn() }));

import * as babel from 'babel-core';
import path from 'path';
import extractStyles from '../extractStyles';
import dedent from 'dedent';

const transpile = source => {
  return babel.transform(source, {
    presets: ['es2015'],
    plugins: [path.resolve('src/babel/index.js')],
    babelrc: false,
  });
};

describe('babel plugin', () => {
  it('should not process tagged template if tag is not `css`', () => {
    const { code } = transpile(dedent`
    const header = \`
      font-size: 3em;
    \`;
    `);
    expect(code.includes('font-size: 3em;')).toBeTruthy();
    expect(code).toMatchSnapshot();
  });

  it('should fail if there are any unresolvable expressions in tagged template', () => {
    expect(() => {
      transpile(
        dedent(`
        const header = css\`
          font-size: \${someFn()}em;
        \`;
        `)
      );
    }).toThrowError(/Cannot resolve expression/);
  });

  it('should generate hash and inject it into `css.named` call for simple fixture', () => {
    const { code } = transpile(dedent`
    const header = css\`
      font-size: 3em;
    \`;
    `);
    expect(code.includes('header_')).toBeTruthy();
    expect(code).toMatchSnapshot();
    // Check if generated hash is the same
    expect(
      transpile(
        dedent(`
        const header = css\`
          font-size: 3em;
        \`;
        `)
      ).code
    ).toMatchSnapshot();
  });

  it('should generate hash and inject it into `css.named` call for multiline fixture', () => {
    const { code } = transpile(dedent`
    const header = css\`
      font-size: 3em;
      color: red;
      border: 1px solid #000;
    \`;
    `);
    expect(code.includes('header_')).toBeTruthy();
    expect(code).toMatchSnapshot();
    // Check if generated hash is the same
    expect(
      transpile(
        dedent(`
        const header = css\`
          font-size: 3em;
          color: red;
          border: 1px solid #000;
        \`;
        `)
      ).code
    ).toMatchSnapshot();
  });

  it('should resolve expressions and generate hash (CommonJS)', () => {
    extractStyles.mockImplementationOnce(tagExpr => {
      expect(tagExpr.quasi.expressions.length).toBe(0);
      expect(
        tagExpr.quasi.quasis[0].value.cooked.includes('font-size: 14px')
      ).toBeTruthy();
    });
    const { code } = transpile(
      dedent(`
      const constants = require('./src/babel/__tests__/__fixtures__/constants.cjs.js');

      const header = css\`
        font-size: \${constants.fontSize};
        color: black;
      \`;
      `)
    );
    expect(code).toMatchSnapshot();
  });

  it('should resolve expressions and generate hash (ESM)', () => {
    extractStyles.mockImplementationOnce(tagExpr => {
      expect(tagExpr.quasi.expressions.length).toBe(0);
      expect(
        tagExpr.quasi.quasis[0].value.cooked.includes('font-size: 14px')
      ).toBeTruthy();
    });
    const { code } = transpile(
      dedent(`
      import constants from './src/babel/__tests__/__fixtures__/constants.esm.js';

      const header = css\`
        font-size: \${constants.fontSize};
        color: black;
      \`;
      `)
    );
    expect(code).toMatchSnapshot();
  });
});
