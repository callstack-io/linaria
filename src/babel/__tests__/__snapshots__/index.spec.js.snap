// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel plugin should generate hash and inject it into \`css.named\` call for multiline fixture 1`] = `
"\\"use strict\\";

var _templateObject = _taggedTemplateLiteral([\\"\\\\n  font-size: 3em;\\\\n  color: red;\\\\n  border: 1px solid #000;\\\\n\\"], [\\"\\\\n  font-size: 3em;\\\\n  color: red;\\\\n  border: 1px solid #000;\\\\n\\"]);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var header = css.named(\\"header_b25fc411\\")(_templateObject);"
`;

exports[`babel plugin should generate hash and inject it into \`css.named\` call for multiline fixture 2`] = `
"\\"use strict\\";

var _templateObject = _taggedTemplateLiteral([\\"\\\\n  font-size: 3em;\\\\n  color: red;\\\\n  border: 1px solid #000;\\\\n\\"], [\\"\\\\n  font-size: 3em;\\\\n  color: red;\\\\n  border: 1px solid #000;\\\\n\\"]);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var header = css.named(\\"header_b25fc411\\")(_templateObject);"
`;

exports[`babel plugin should generate hash and inject it into \`css.named\` call for simple fixture 1`] = `
"\\"use strict\\";

var _templateObject = _taggedTemplateLiteral([\\"\\\\n  font-size: 3em;\\\\n\\"], [\\"\\\\n  font-size: 3em;\\\\n\\"]);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var header = css.named(\\"header_df78b444\\")(_templateObject);"
`;

exports[`babel plugin should generate hash and inject it into \`css.named\` call for simple fixture 2`] = `
"\\"use strict\\";

var _templateObject = _taggedTemplateLiteral([\\"\\\\n  font-size: 3em;\\\\n\\"], [\\"\\\\n  font-size: 3em;\\\\n\\"]);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var header = css.named(\\"header_df78b444\\")(_templateObject);"
`;

exports[`babel plugin should not process tagged template if tag is not \`css\` 1`] = `
"\\"use strict\\";

var header = \\"\\\\n  font-size: 3em;\\\\n\\";"
`;

exports[`babel plugin should resolve expressions and generate hash (CommonJS) 1`] = `
"'use strict';

var _templateObject = _taggedTemplateLiteral(['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n'], ['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var constants = require('./src/babel/__tests__/__fixtures__/constants.cjs.js');

var header = css.named('header_a6d91e89')(_templateObject);"
`;

exports[`babel plugin should resolve expressions and generate hash (ESM) 1`] = `
"'use strict';

var _templateObject = _taggedTemplateLiteral(['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n'], ['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n']);

var _constantsEsm = require('./src/babel/__tests__/__fixtures__/constants.esm.js');

var _constantsEsm2 = _interopRequireDefault(_constantsEsm);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var header = css.named('header_a6d91e89')(_templateObject);"
`;

exports[`babel plugin should resolve expressions from an object with constants 1`] = `
"'use strict';

var _templateObject = _taggedTemplateLiteral(['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n'], ['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var constants = { fontSize: '14px', color: 'black' };

var header = css.named('header_a6d91e89')(_templateObject);"
`;

exports[`babel plugin should resolve expressions from an object with constants with advanced destructurization 1`] = `
"'use strict';

var _templateObject = _taggedTemplateLiteral(['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n'], ['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var _color = { color: 'black' },
    _color$fontSize = _color.fontSize,
    fontSize = _color$fontSize === undefined ? '14px' : _color$fontSize,
    fontColor = _color.color;


var header = css.named('header_a6d91e89')(_templateObject);"
`;

exports[`babel plugin should resolve expressions from an object with constants with destructurization 1`] = `
"'use strict';

var _templateObject = _taggedTemplateLiteral(['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n'], ['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var _fontSize$color = { fontSize: '14px', color: 'black' },
    fontSize = _fontSize$color.fontSize,
    color = _fontSize$color.color;


var header = css.named('header_a6d91e89')(_templateObject);"
`;

exports[`babel plugin should resolve multiple expressions 1`] = `
"'use strict';

var _templateObject = _taggedTemplateLiteral(['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n  background-color: black;\\\\n  font-weight: bold;\\\\n'], ['\\\\n  font-size: 14px;\\\\n  color: black;\\\\n  background-color: black;\\\\n  font-weight: bold;\\\\n']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var constants = {
  color: 'black',
  fontSize: '14px',
  backgroundColor: 'black',
  fontWeight: 'bold'
};

var header = css.named('header_2add690b')(_templateObject);"
`;
